src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                        // even if the CompletionService returns item for a local variable "task" which is an exact match.
+                        var itemWithMatch = items[i];
+                        if (itemWithMatch.RoslynCompletionItem.DisplayText != defaultText)
+                            continue;
+
+                        if (itemWithMatch.PatternMatch != null && itemWithMatch.PatternMatch.Value.Kind > PatternMatchKind.Prefix)
+                            break;
+
+                        // Always hard-select except in suggestion mode.
+                        return _hasSuggestedItemOptions
+                            ? intialSelection with { SelectedItemIndex = i }
+                            : intialSelection with { SelectedItemIndex = i, SelectionHint = UpdateSelectionHint.Selected };
+                    }
+                }
+
+                return intialSelection;
+            }
+
+            /// <summary>
+            /// Compare the pattern matching result of the current selection with the pattern matching result of the suggested defaults (both w.r.t. the filter text.)
+            /// If the suggested default is no worse than current selected item (in a case-sensitive manner,) use the suggested default. Otherwise use the original selection.
+            /// For example, if user typed "C", roslyn might select "CancellationToken", but with suggested default "Console" we will end up selecting "Console" instead.
+            /// </summary>
+            private ItemSelection GetDefaultsMatch(IReadOnlyList<MatchResult<VSCompletionItem>> items, ItemSelection intialSelection)
+            {
+                // Because the items are already sorted based on pattern-matching score, try to limit the range for the items we compare default with
+                // by searching for the first "inferior" item, so we can avoid always going through the entire list.
+                int inferiorItemIndex;
+                if (_filterText.Length == 0)
+                {
+                    // Without filterText, all items are equally good match (w.r.t to the empty filterText), so we have to consider all of them.
+                    inferiorItemIndex = items.Count;
+                }
+                else
+                {
+                    // Because the items are sorted based on pattern-matching score, the selectedIndex is in the middle of a range of
+                    // -- as far as the pattern matcher is concerned -- equivalent items (items with identical PatternMatch.Kind and IsCaseSensitive).
+                    // Find the last items in the range and use that to limit the items searched for from the defaults list.          
+                    var selectedItemMatch = items[intialSelection.SelectedItemIndex].PatternMatch;
+
+                    if (!selectedItemMatch.HasValue)
+                        return intialSelection;
```

Comment:
we need docs explaining what's going on.  i highly recommend (for lots of htese methods and these early returns and whatnot, giving a concrete case of what the item was, and why we're now bailing).